libs/ui/src/images/normalize.ts

```diff
@@ -0,0 +1,161 @@
+  | { code: 'unsupportedImageType' }
+  | { code: 'unexpected'; error: unknown };
+
+export type NormalizeParams = {
+  minHeightPx?: number;
+  minWidthPx?: number;
+} & (
+  | { maxHeightPx: number; maxWidthPx?: number }
+  | { maxHeightPx?: number; maxWidthPx: number }
+);
+
+/**
+ * Normalizes an image, potentially resizing it to fit within the given bounds,
+ * and returns a serialized data URL version of the final image with the same
+ * encoding as the original.
+ *
+ * If minimum dimensions are specified, an error is returned for images that
+ * fall below the given thresholds (see {@link NormlaizeError}).
+ */
+export async function normalizeFile(
+  image: File,
+  params: NormalizeParams
+): Promise<NormalizeResult> {
+  const mimeType = image.type;
+  if (!isSupportedMimeType(mimeType)) {
+    return err({ code: 'unsupportedImageType' });
+  }
+
+  return normalizeDataUrl(await serialize(image), mimeType, params);
+}
+
+const SUPPORTED_MIME_TYPES = ['image/png', 'image/jpeg'] as const;
+export type ImageType = (typeof SUPPORTED_MIME_TYPES)[number];
+
+/**
+ * Normalizes an image, potentially resizing it to fit within the given bounds,
+ * and returns a serialized data URL version of the final image with the same
+ * encoding as the original.
+ *
+ * If minimum dimensions are specified, an error is returned for images that
+ * fall below the given thresholds (see {@link NormlaizeError}).
+ */
+export async function normalizeDataUrl(
+  imageDataUrl: string,
+  mimeType: ImageType,
+  params: NormalizeParams
+): Promise<NormalizeResult> {
+  const { promise, resolve } = deferred<NormalizeResult>();
+
+  const img = new Image();
+  async function onLoad() {
+    if (params.minHeightPx && img.height < params.minHeightPx) {
+      resolve(err({ code: 'belowMinHeight', heightPx: img.height }));
+    }
+
+    if (params.minWidthPx && img.width < params.minWidthPx) {
+      resolve(err({ code: 'belowMinWidth', widthPx: img.width }));
+    }
+
+    try {
+      const scaleX = (params.maxWidthPx || img.width) / img.width;
+      const scaleY = (params.maxHeightPx || img.height) / img.height;
+      const scale = Math.min(scaleX, scaleY);
+
+      if (scale >= 1) {
+        return resolve(
+          ok({
+            dataUrl: imageDataUrl,
+            heightPx: img.height,
+            widthPx: img.width,
+          })
+        );
+      }
+
+      const canvas = document.createElement('canvas');
+      const context = assertDefined(canvas.getContext('2d'));
+
+      canvas.width = img.width * scale;
+      canvas.height = img.height * scale;
+      context.drawImage(img, 0, 0, canvas.width, canvas.height);
+
+      resolve(
+        ok({
+          dataUrl: await canvasToDataUrl(canvas, mimeType),
+          heightPx: canvas.height,
+          widthPx: canvas.width,
+        })
+      );
+    } catch (error) {
+      resolve(err({ code: 'unexpected', error }));
+    }
+  }
+
+  img.addEventListener('load', onLoad, { once: true });
```

Comment:
Does this image and event listener get garbage collected? Or do we need to manually dispose?